docs/resources/server/recipes/custom.md

```diff
@@ -0,0 +1,134 @@
+Custom Recipes
+==============
+
+Every recipe definition is made up of three components: the `Recipe` implementation which holds the data and handles the execution logic with the provided inputs, the `RecipeType` which represents the category or context the recipe will be used in, and the `RecipeSerializer` which handles decoding and network communication of the recipe data. How one chooses to use the recipe is up to the implementor.
+
+Recipe
+------
+
+The `Recipe` interface describes the recipe data and the execution logic. This includes matching the inputs and providing the associated result. As the recipe subsystem performs item transformations by default, the inputs are supplied through a `Container` subtype.
+
+!!! important
+    The `Container`s passed into the recipe should be treated as if its contents were immutable. Any mutable operations should be performed on a copy of the input through `ItemStack#copy`.
+
+To be able to obtain a recipe instance from the manager, `#matches` must return true. This method checks against the provided container to see whether the associated inputs are valid. `Ingredient`s can be used for validation by calling `Ingredient#test`.
+
+If the recipe has been chosen, it is then built using `#assemble` which may use data from the inputs to create the result.
+
+!!! tip
+    `#assemble` should always produce a unique `ItemStack`. If unsure whether `#assemble` does so, call `ItemStack#copy` on the result before returning.
+
+Most of the other methods are purely for integration with the recipe book.
+
+```java
+public record ExampleRecipe(Ingredient input, int data, ItemStack output) implements Recipe<Container> {
+  // Implement methods here
+}
+```
+
+RecipeType
+----------
+
+`RecipeType` is responsible for defining the category or context the recipe will be used within. For example, if a recipe was going to be smelted in a furnace, it would have a type of `RecipeType#SMELTING`.  Being blasted in a blast furnace would have a type of `RecipeType#BLASTING`.
```

Comment:
Extra whitespace:
```suggestion
`RecipeType` is responsible for defining the category or context the recipe will be used within. For example, if a recipe was going to be smelted in a furnace, it would have a type of `RecipeType#SMELTING`. Being blasted in a blast furnace would have a type of `RecipeType#BLASTING`.
```